server/sonar-server/src/test/java/org/sonar/server/edition/ws/ApplyLicenseActionTest.java

```diff
@@ -125,25 +142,109 @@ public void request_fails_with_BadRequestException_is_pendingStatus_is_not_NONE(
   }
 
+  @Test
+  public void apply_without_need_to_install() throws IOException {
+    userSessionRule.logIn().setSystemAdministrator();
+    setPendingLicense(NONE);
+    when(editionInstaller.requiresInstallationChange(Collections.singleton("plugin1"))).thenReturn(false);
+
+    TestRequest request = actionTester.newRequest()
+      .setMediaType(MediaTypes.PROTOBUF)
+      .setParam(PARAM_LICENSE, createLicenseParam(PENDING_EDITION_NAME, "plugin1"));
+
+    TestResponse response = request.execute();
+    assertResponse(response, PENDING_EDITION_NAME, "", NONE);
+    verify(mutableEditionManagementState).newEditionWithoutInstall(PENDING_EDITION_NAME);
+  }
+
+  @Test
+  public void apply_offline() throws IOException {
+    userSessionRule.logIn().setSystemAdministrator();
+    setPendingLicense(PendingStatus.MANUAL_IN_PROGRESS);
+    when(editionInstaller.requiresInstallationChange(Collections.singleton("plugin1"))).thenReturn(true);
+    when(editionInstaller.install(Collections.singleton("plugin1"))).thenReturn(false);
+
+    TestRequest request = actionTester.newRequest()
+      .setMediaType(MediaTypes.PROTOBUF)
+      .setParam(PARAM_LICENSE, createLicenseParam(PENDING_EDITION_NAME, "plugin1"));
+
+    TestResponse response = request.execute();
+
+    assertResponse(response, PENDING_EDITION_NAME, "", PendingStatus.MANUAL_IN_PROGRESS);
+    verify(mutableEditionManagementState).startManualInstall(any(License.class));
+  }
+
+  @Test
+  public void apply_successfully_auto_installation() throws IOException {
+    userSessionRule.logIn().setSystemAdministrator();
+    setPendingLicense(PendingStatus.AUTOMATIC_IN_PROGRESS);
+    when(editionInstaller.requiresInstallationChange(Collections.singleton("plugin1"))).thenReturn(true);
+    when(editionInstaller.install(Collections.singleton("plugin1"))).thenReturn(true);
+
+    TestRequest request = actionTester.newRequest()
+      .setMediaType(MediaTypes.PROTOBUF)
+      .setParam(PARAM_LICENSE, createLicenseParam(PENDING_EDITION_NAME, "plugin1"));
+
+    TestResponse response = request.execute();
+
+    assertResponse(response, PENDING_EDITION_NAME, "", PendingStatus.AUTOMATIC_IN_PROGRESS);
+    verify(mutableEditionManagementState).startAutomaticInstall(any(License.class));
+  }
+
+  private void assertResponse(TestResponse response, String expectedNextEditionKey, String expectedEditionKey,
+    PendingStatus expectedPendingStatus) throws IOException {
+    StatusResponse parsedResponse = WsEditions.StatusResponse.parseFrom(response.getInputStream());
+    assertThat(parsedResponse.getCurrentEditionKey()).isEqualTo(expectedEditionKey);
+    assertThat(parsedResponse.getNextEditionKey()).isEqualTo(expectedNextEditionKey);
+    assertThat(parsedResponse.getInstallationStatus()).isEqualTo(WsEditions.InstallationStatus.valueOf(expectedPendingStatus.toString()));
+  }
+
+  private void setPendingLicense(PendingStatus pendingStatus) {
```

Comment:
hum... this works, but it's highly misleading as it mocks states which the production implementation can't get to.

I wonder if it would be safer to use the production implementation of `MutableEditionManagementState` in the test. A bit less convenient, but ensures relevant tests.